tests/conftest.py

```diff
@@ -723,25 +741,45 @@ def setup_recorder(config=None):
 
 
 @pytest.fixture
-async def recorder_mock(enable_nightly_purge, enable_statistics, hass):
-    """Fixture with in-memory recorder."""
-    stats = recorder.Recorder.async_periodic_statistics if enable_statistics else None
-    nightly = recorder.Recorder.async_nightly_tasks if enable_nightly_purge else None
-    with patch(
-        "homeassistant.components.recorder.Recorder.async_periodic_statistics",
-        side_effect=stats,
-        autospec=True,
-    ), patch(
-        "homeassistant.components.recorder.Recorder.async_nightly_tasks",
-        side_effect=nightly,
-        autospec=True,
-    ):
-        await async_init_recorder_component(hass)
-        await hass.async_start()
-        await hass.async_block_till_done()
-        await hass.async_add_executor_job(
-            hass.data[recorder.DATA_INSTANCE].block_till_done
+async def async_setup_recorder_instance(
+    enable_nightly_purge, enable_statistics
+) -> AsyncGenerator[SetupRecorderInstanceT, None]:
+    """Yield callable to setup recorder instance."""
+
+    async def async_setup_recorder(
+        hass: HomeAssistant, config: ConfigType | None = None
+    ) -> recorder.Recorder:
+        """Setup and return recorder instance."""  # noqa: D401
+        nightly = (
+            recorder.Recorder.async_nightly_tasks if enable_nightly_purge else None
         )
+        stats = (
+            recorder.Recorder.async_periodic_statistics if enable_statistics else None
+        )
+        with patch(
+            "homeassistant.components.recorder.Recorder.async_nightly_tasks",
+            side_effect=nightly,
+            autospec=True,
+        ), patch(
+            "homeassistant.components.recorder.Recorder.async_periodic_statistics",
+            side_effect=stats,
+            autospec=True,
+        ):
+            await async_init_recorder_component(hass, config)
+            await hass.async_block_till_done()
+            instance = hass.data[recorder.DATA_INSTANCE]
+            # The recorder's worker is not started until Home Assistant is running
+            if hass.state == CoreState.running:
+                await async_recorder_block_till_done(hass, instance)
+            return instance
+
+    yield async_setup_recorder
```

Comment:
you can just return it since you don't do anything afterwards
```suggestion
    return async_setup_recorder
```